pkg/ipcache/types.go

```diff
@@ -186,208 +245,207 @@ func (s prefixInfo) sortedBySourceThenResourceID() []ipcachetypes.ResourceID {
-
-		tunnelPeer           ipcachetypes.TunnelPeer
-		tunnelPeerResourceID ipcachetypes.ResourceID
-
-		encryptKey           ipcachetypes.EncryptKey
-		encryptKeyResourceID ipcachetypes.ResourceID
-
-		requestedID           ipcachetypes.RequestedIdentity
-		requestedIDResourceID ipcachetypes.ResourceID
-
-		endpointFlags           ipcachetypes.EndpointFlags
+		overrideResourceID      ipcachetypes.ResourceID
+		tunnelPeerResourceID    ipcachetypes.ResourceID
+		encryptKeyResourceID    ipcachetypes.ResourceID
+		requestedIDResourceID   ipcachetypes.ResourceID
 		endpointFlagsResourceID ipcachetypes.ResourceID
 	)
 
+	labelResourceIDs := map[string]ipcachetypes.ResourceID{}
+
 	for _, resourceID := range s.sortedBySourceThenResourceID() {
-		info := s[resourceID]
+		info := s.byResource[resourceID]
 
-		if info.identityOverride {
-			if len(override) > 0 {
-				scopedLog.Warn(
-					"Detected conflicting identity override for prefix. "+
-						"This may cause connectivity issues for this address.",
-					logfields.Identity, override,
-					logfields.Resource, overrideResourceID,
-					logfields.ConflictingIdentity, info.labels,
-					logfields.ConflictingResource, resourceID,
-				)
+		// Sorted by source priority, so the first source wins.
+		if out.source == "" {
+			out.source = info.source
+		}
+
+		if len(info.labels) > 0 {
+			if len(out.labels) > 0 {
+				// merge labels, complaining if the value exists
+				for key, newLabel := range info.labels {
+					otherLabel, exists := out.labels[key]
+					if exists && !otherLabel.DeepEqual(&newLabel) {
+						scopedLog.Warn(
+							"Detected conflicting label for prefix. "+
+								"This may cause connectivity issues for this address.",
+							logfields.Labels, out.labels,
+							logfields.Resource, labelResourceIDs[key],
+							logfields.ConflictingLabels, otherLabel,
+						)
+					} else if !exists {
+						out.labels[key] = newLabel
+						labelResourceIDs[key] = resourceID
+					}
+				}
+			} else {
+				out.labels = labels.NewFrom(info.labels) // copy map, as we will be mutating it
 			}
+		}
 
+		if info.identityOverride {
 			if len(info.labels) == 0 {
 				scopedLog.Warn(
 					"Detected identity override, but no labels where specified. "+
 						"Falling back on the old non-override labels. "+
 						"This may cause connectivity issues for this address.",
 					logfields.Resource, resourceID,
-					logfields.IdentityOld, s.ToLabels(),
 				)
 			} else {
-				override = info.labels
-				overrideResourceID = resourceID
+				if out.identityOverride {
+					scopedLog.Warn(
+						"Detected conflicting identity override for prefix. "+
+							"This may cause connectivity issues for this address.",
+						logfields.Labels, out.labels,
+						logfields.Resource, overrideResourceID,
+						logfields.ConflictingLabels, info.labels,
+						logfields.ConflictingResource, resourceID,
+					)
+				} else {
+					out.identityOverride = true
+					out.labels = info.labels
```

Comment:
If we have an identity override, we must not merge any additional labels into `out.labels`. I think this is no longer the case with the new implementation, since we'll continue to merge the next `info.labels` into `out.labels` without checking if `out.identityOverride` is set.

Reply:
This makes me wonder if we have a cilium-cli CI test case that uses this.

Reply:
I don't think we do. The original user was/is external workloads, for which we removed support in the Cilium CLI a while a go. The code in Cilium CNI has been deprecated but not yet removed to my knowledge.

Apparently we also use the feature in the code below, maybe for that particular case there's another solution, so we can get rid of this override function completely

https://github.com/cilium/cilium/blob/a719911acb08f86a815159ee296b9046b45a767d/daemon/cmd/hostips-sync.go#L202

Reply:
Ah, so apparently external workload has been removed: https://github.com/cilium/cilium/pull/37418

But the code overwriting the identity based on node labels hasn't.